### DISCUSSION TITLE
Error writing to "config/params/generated_config.php

### DISCUSSION THREAD

**Discussion by franco640** [shlinkio/discussions#142] [Help wanted] [answered]
### Shlink version

shlink4.0.3_php8.3_dist

### PHP version

PHP 8.3.4

### How do you serve Shlink

Self-hosted Apache

### Database engine

MariaDB

### Database version

10.6.16-MariaDB-0 ubuntu0.22.04.1 Ubuntu 22.04

### Current behavior

When i finish the installer prompts and is trying to write to the generated_config.php I get an error 

In PhpArray.php line 118:

  Error writing to "config/params/generated_config.php": file_put_contents(config/params/generated_config.php): Failed to open stream: No such file or directory

### Expected behavior

It should create and write to that php file as mentioned above, 

### Minimum steps to reproduce

doing the basic user setup I did the following 
installed the following  

sudo apt install ca-certificates apt-transport-https software-properties-common lsb-release -y
sudo add-apt-repository ppa:ondrej/php -y

sudo apt install php8.3 libapache2-mod-php8.3 php8.3-curl php8.3-intl php8.3-gd php8.3-common php8.3-fpm php8.3-common php8.3-mysql php8.3-xml php8.3-apcu -y

enable 
sudo a2dismod mpm_prefork php8.3
sudo a2enmod mpm_event rewrite proxy_fcgi setenvif php8.3
sudo a2enconf php8.3-fpm

Maria DB

1 - sudo apt install mariadb-server mariadb-client
2 - sudo mysql_secure_installation
3 - sudo mariadb -u root
4 - (Test credentials ) 
         CREATE DATABASE shlink;
         GRANT ALL ON shlink.* to 'shlink'@'localhost' IDENTIFIED BY 'password!';
         FLUSH PRIVILEGES;
         EXIT;

Download dist file 

wget https://github.com/shlinkio/shlink/releases/download/v4.0.3/shlink4.0.3_php8.3_dist.zip
sudo unzip shlink4.0.3_php8.3_dist.zip -d /var/www
sudo mv /var/www/shlink* /var/www/shlink
sudo chown -R www-data:www-data /var/www/shlink
chmod -R 755 /var/www/shlink
sudo -u www-data php8.3 /var/www/shlink/vendor/shlinkio/shlink-installer/bin/shlink-installer install

Options chosen 

 [1] Classic web server (Nginx, Apache, etc)
 [1] MariaDB

 Database name [shlink]:
 > shlink

Database host [localhost]:
 >[enter]

 Database port [3306]:
 >[enter]

 Database username:
 > shlink

 Database password:
 >password - this is different to mine of course 

 Unix socket (leave empty to not use a socket):
 >[enter]

 Default domain for generated short URLs:
 >(enter my url shortner domain)

 Is HTTPS enabled on this server? (yes/no) [yes]:
 >[enter]

 What is the default length you want generated short codes to have? (You will still be able to override this on every created short URL) [5]:
 >[enter]

 Do you want Shlink to resolve the short URL title based on the long URL's title tag (if any)? Otherwise, it will be kept empty unless explicitly provided. (yes/no) [yes]:
 >[enter]

 Do you want Shlink to redirect short URLs as soon as the first segment of the path matches a short code, appending the rest to the long URL?

 >[enter]

 Do you want to support short URLs with multi-segment custom slugs? (for example, https://example.com/foo/bar) [Slugs and short codes will support only one segment (https://example.com/foo). Orphan visits will have one of "base_url", 

>yes

Do you want to support trailing slashes in short URLs? (https://s.test/foo and https://s.test/foo/ will be considered the same) (yes/no) [no]:
 >yes

 How do you want short URLs to be matched?
 Warning! This feature is experimental. It only applies to public routes (short URLs and QR codes). REST API routes always use strict match.

>loose

 Provide a GeoLite2 license key. Leave empty to disable geolocation. (Go to https://shlink.io/documentation/geolite-license-key to know how to generate it):
 > (put my key here) 

 What kind of redirect do you want your short URLs to have? [All visits will always be tracked. Not that good for SEO. Only GET requests will be redirected.]:

 >308

 How long (in seconds) do you want your redirects to be cached by visitors? [30]:
 > Enter

TRACKING
========

I left it all default 

REDIRECTS
=========

used my domain 

QR CODES
========

Used defaults

APPLICATION
===========

 What is the path from which shlink is going to be served? (It must include a leading bar, like "/shlink". Leave empty if you plan to serve shlink from the root of the domain):
 >/shlink ( i used multiple paths including no /shlink with teh same results)

Set the timezone in which your Shlink instance is running (leave empty to use the one set in PHP config):
 >UTC

INTEGRATIONS
============

Left defaults 

then get  

Error writing to "config/params/generated_config.php": file_put_contents(config/params/generated_config.php):

This is very fustrating and have spent several days trying to get this to work

**Comment by acelaya** [comment]
Using `sudo` all over is likely going to cause problems down the line.

I would recommend not running all the commands under `Download dist file` with `sudo`.

**Reply by franco640** [reply]
Same issue occurs, i have done this so many times and it's very frustrating.

**Reply by acelaya** [reply]
The error is caused by missing write permissions.

**Reply by franco640** [reply]
what permissions are required ? on the params folder i have  755 - drwxr-xr-x  www-data www-data  params

**Reply by franco640** [reply]
also the config folder is 755 - drwxr-xr-x   www-data www-data.

**Comment by franco640** [comment]
base path for the Shlink folder - Shortner-Server:/var/www/shlinkio$ cd /var/www/shlinkio/

I can do a video of the settings i'm using

**Comment by franco640** [comment, marked as answer]
I have it working finally. 

I did not know that I had to be in the root folder of where you extracted the dist file. For myself it was 

cd /var/www/shlinkio 

and then run 

sudo -u www-data php8.3 /var/www/shlinkio/vendor/shlinkio/shlink-installer/bin/shlink-installer install

for it to work 

not great and caused me no end of grief to get this to work.

**Reply by acelaya** [reply]
Yep, that's documented here https://shlink.io/documentation/command-line-interface/installation-tool/

![Screenshot_20240326-113522.png](https://github.com/shlinkio/shlink/assets/2719332/a85ec65e-caa8-47ab-ac5a-a49c5aea4570)

**Reply by franco640** [reply]
bugger, my bad. That's annoying

**Reply by franco640** [reply]
Was looking at the youtube videos when i was doing the installation. They all seem to be very outdated
